From f907ec5099f610445852fad3127b479a068eaee0 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:11:47 -0400
Subject: [alsa] hda: Explicitly keep codec powered up in hdmi_present_sense

Message-id: <140972823573-4-patchset-jkysela@redhat.com>
Patchwork-id: 90704
O-Subject: [RHEL 7.1 PATCH 003/442] ALSA: hda - Explicitly keep codec powered up in hdmi_present_sense
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

This should help us avoid the following mutex deadlock:

[] mutex_lock+0x2a/0x50
[] hdmi_present_sense+0x53/0x3a0 [snd_hda_codec_hdmi]
[] generic_hdmi_resume+0x5a/0x70 [snd_hda_codec_hdmi]
[] hda_call_codec_resume+0xec/0x1d0 [snd_hda_codec]
[] snd_hda_power_save+0x1e4/0x280 [snd_hda_codec]
[] codec_exec_verb+0x5f/0x290 [snd_hda_codec]
[] snd_hda_codec_read+0x5b/0x90 [snd_hda_codec]
[] snd_hdmi_get_eld_size+0x1e/0x20 [snd_hda_codec_hdmi]
[] snd_hdmi_get_eld+0x2c/0xd0 [snd_hda_codec_hdmi]
[] hdmi_present_sense+0x9a/0x3a0 [snd_hda_codec_hdmi]
[] hdmi_repoll_eld+0x34/0x50 [snd_hda_codec_hdmi]

Signed-off-by: David Henningsson <david.henningsson@canonical.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: da4a7a3926d09c13ae052ede67feb7285e01e3f5)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_hdmi.c b/sound/pci/hda/patch_hdmi.c
index f5060fc..977db17 100644
--- a/sound/pci/hda/patch_hdmi.c
+++ b/sound/pci/hda/patch_hdmi.c
@@ -1496,11 +1496,14 @@ static bool hdmi_present_sense(struct hdmi_spec_per_pin *per_pin, int repoll)
   * specification worked this way. Hence, we just ignore the data in
   * the unsolicited response to avoid custom WARs.
   */
- int present = snd_hda_pin_sense(codec, pin_nid);
+ int present;
  bool update_eld = false;
  bool eld_changed = false;
  bool ret;
 
+ snd_hda_power_up(codec);
+ present = snd_hda_pin_sense(codec, pin_nid);
+
  mutex_lock(&per_pin->lock);
  pin_eld->monitor_present = !!(present & AC_PINSENSE_PRESENCE);
  if (pin_eld->monitor_present)
@@ -1573,6 +1576,7 @@ static bool hdmi_present_sense(struct hdmi_spec_per_pin *per_pin, int repoll)
   jack->block_report = !ret;
 
  mutex_unlock(&per_pin->lock);
+ snd_hda_power_down(codec);
  return ret;
 }
 
-- 
1.7.1